metagpt/actions/design.py

```diff
@@ -0,0 +1,127 @@
+import json
+from functools import wraps
+from typing import Callable, Any, List, Optional, Tuple
+
+# Local library imports
+from metagpt.actions import Action
+from metagpt.logs import logger
+from metagpt.utils.common import OutputParser
+from metagpt.prompts.sd_design import (
+    SD_PROMPT_KW_OPTIMIZE_TEMPLATE,
+    SD_PROMPT_IMPROVE_OPTIMIZE_TEMPLATE,
+    FORMAT_INSTRUCTIONS,
+    PROMPT_OUTPUT_MAPPING
+)
+
+
+class Tool:
+    """Define a tool with its name, function and description."""
+    
+    def __init__(self, name: str, func: Callable, description: str) -> None:
+        """Initialize tool."""
+        self.name = name
+        self.func = func
+        self.description = description
+
+# Decorator for the BaseModelAction to wrap it with system primer details
+def system_primer_decorator(func):
+    @wraps(func)
+    async def wrapper(self, *args, **kwargs):
+        system_primer = self.SYSTEM_PRIMER_TEMPLATE.format(answer_count=kwargs.get('answer_count', 1))
+        logger.info(system_primer)
+        return await func(self, *args, system_primer=system_primer, **kwargs)
+    
+    return wrapper
+
+
+class BaseModelAction(Action):
+    # A default template for the system primer.
+    SYSTEM_PRIMER_TEMPLATE = "Act like you are a terminal and always format your response as json. Always return exactly {answer_count} answers per question."
+    
+    def __init__(self, name: str = "", description: str = "", *args, **kwargs):
+        super().__init__(name, *args, **kwargs)
+        self.desc = description
+    
+    async def handle_response(self, resp: str, key: str) -> Any:
+        """Handle JSON response and extract value by a given key."""
+        try:
+            resp_json = json.loads(resp)
+            return resp_json[key]
+        except json.JSONDecodeError:
+            logger.error("Invalid JSON response!")
+            return None
+        except KeyError:
+            logger.error(f"Key '{key}' not found in the JSON response!")
+            return None
+    
+    @system_primer_decorator
+    async def run_optimize_or_improve(self, query: str, domain: str, template: str, answer_count: int = 1,
+                                      system_primer=None) -> List[str]:
+        """Run optimization or improvement based on the given template."""
+        prompt = template.format(messages=query, domain=domain, answer_count=answer_count)
+        resp: str = await self._aask(prompt=prompt, system_msgs=[system_primer])
+        result = await self.handle_response(resp, key="responses")
+        return result or [query]
+
+
+class SDPromptOptimize(BaseModelAction):
+    """
+    Optimize graphical prompts based on keywords.
+    扩充画图的提示词，根据keyword
+    """
+    
+    def __init__(self, name: str = "", *args, **kwargs):
+        super().__init__(name, description="PromptOptimize", *args, **kwargs)
+    
+    async def run(self, query: str, domain: str = "realistic", answer_count: int = 1) -> List[str]:
```

Comment:
Does `domain` has a suggested set? If it's yes, suggest use Enum.

Reply:
At the moment, there are only options for realistic and anime/gaming styles.

Reply:
Yeah. Where can users find the options set? Maybe put it in the param description or in other ways.